### PR TITLE
Add pagination and summary to Disputes table

### LINKED ISSUE #3386: Disputes: add pagination functionality
See p1636762622030300-slack-CGGCLBN58

Similar to transactions and deposits, disputes need pagination functionality.

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #3386 

#### Changes proposed in this Pull Request

Adds summary information and pagination controls to the Disputes table. 

Screen Shot 2021-12-26 at 19 20 18

#### Testing instructions

1. Add a number of orders with disputes greater than the rows per page defined for the disputes table. This number is set to 25. Alternatively, you can use the `per_page` query parameter to set a smaller number i.e. `http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fdisputes&paged=1&per_page=10`
2. You can use [Stripe's test cards to pay for the orders](https://stripe.com/docs/testing#disputes). Using one of this cards will generate a dispute after the order is paid.
3. Note that this PR relies on the pagination functionality developed [here](1415-gh-automattic/woocommerce-payments-server)
4. Follow instructions from 1408-gh-Automattic/woocommerce-payments-server to have disputes saved in the cache
5. Observe that the pagination controls appear at the bottom of the table.
6. Check that the pagination controls work as expected (navigate to next/ prev page, go to page number, select rows per page)
7. Observe that the total number of Disputes is displayed at the bottom of the table

-------------------

- [x] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (kalessil) [inline includes/wc-payment-api/class-wc-payments-api-client.php:697]: I believe we can make the parameters/return type strictly typed here.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (kalessil) [inline includes/wc-payment-api/class-wc-payments-api-client.php:728]: I believe we can make the parameters/return type strictly typed here.

File: tests/unit/wc-payment-api/test-class-wc-payments-api-client.php
Comment (kalessil) [inline tests/unit/wc-payment-api/test-class-wc-payments-api-client.php:1323]: That would be awesome, we need better tests coverage in the client repository.

File: includes/admin/class-wc-rest-payments-disputes-controller.php
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-disputes-controller.php:98]: We should doc-block the return type here (like in `get_dispute`). We also probably accept request object as an argument here, or?

File: client/disputes/style.scss
Comment (kalessil) [inline client/disputes/style.scss:7]: Nice!

**Review by kalessil — COMMENTED** [review]
Leaving comments, but overall looks great!

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (mgascam) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Done thanks

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (mgascam) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Done thanks

File: client/disputes/index.tsx
Comment (KarlisJ) [inline client/disputes/index.tsx:110]: Aren't you overriding the previous `isLoading` variable here? I think we should keep hold of both API calls state separately.

File: client/disputes/index.tsx
Comment (KarlisJ) [inline client/disputes/index.tsx:124]: I have multiple concerns here. 

The basis of them is of different structures that the client returns from the server. The cached disputes that are used for listing disputes differ from the proxied dispute that is used for details view. 

Firstly, the type hinting is incorrect. The cached dispute for example doesn't have `evidence_details`. This is going to be quite confusing so I'd suggest specifying in code that for listing disputes we use `CachedDispute` interface. 

Secondly, for the `evidence_details` example - this should not be present from the cached response. `id` also should be `dispute_id`.

Comment (KarlisJ) [reply client/disputes/index.tsx:124]: Ok, I found `format_dispute_for_ui` method that does the conversion in PHP side but there I still have concerns
1. The interface for Dispute is still not complete. To give example, details page takes the single dispute from list that was previously fetched and is missing required fields
2. Not sure if we want to drag the old structure forever into the codebase.

File: client/types/disputes.d.ts
Comment (KarlisJ) [inline client/types/disputes.d.ts:79]: Doesn't this syntax mean that `charge_id`, `source`, etc keys can be absent as not set? In reality, I think server could return empty value, but the key still should be there. 

Or could be I'm misunderstanding something in TS syntax...

File: client/disputes/index.tsx
Comment (dpaun1985) [inline client/disputes/index.tsx:117]: we could use the `dispute` object instead of destructuring it.

File: client/disputes/test/__snapshots__/index.tsx.snap
Comment (dpaun1985) [inline client/disputes/test/__snapshots__/index.tsx.snap:296]: why we have different snapshoots ? shouldn't be the same as before ?

**Review by dpaun1985 — COMMENTED** [review]
Looks good overall! I left some comments.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (KarlisJ) [inline includes/wc-payment-api/class-wc-payments-api-client.php:1768]: Can be deleted.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (KarlisJ) [inline includes/wc-payment-api/class-wc-payments-api-client.php:731]: Why is this block removed? We need this to have a link to order in the UI.

Comment (dmvrtx) [reply includes/wc-payment-api/class-wc-payments-api-client.php:731]: I think it would be more efficient to request dispute data with expanded `charge` object in the back-end, and extract order number from the metadata. `add_order_info_to_object` can be a costly operation.

Comment (KarlisJ) [reply includes/wc-payment-api/class-wc-payments-api-client.php:731]: I see a possibility to improve performance but we can't simply drop this block and add charge in the back-end. Yes, we could remove the query to get the order number but we still need mapping to fetch full order and build the info array that `add_order_info_to_object` does. Back-end doesn't have that information available (nor should it have the responsibility of doing that).

I would also argue the overall performance gain when you take into account necessary back-end changes.

File: client/disputes/index.tsx
Comment (mgascam) [reply client/disputes/index.tsx:112]: No, I don't think it is overridden in this case. The call to `useDisputesSummary` returns an object that has the `isLoading` property, but it is stored into `isSummaryLoading ` variable when the destructuring is done. We have two variables to indicate the loading state: `isLoading` for disputes list and `isSummaryLoading` for disputes summary.

File: client/disputes/index.tsx
Comment (mgascam) [inline client/disputes/index.tsx:115]: I think this is probably better:

```suggestion
	const rows = disputes.map( ( {
			dispute_id,
			amount = 0,
			currency,
			reason,
			source,
			order_number,
			customer_name,
			customer_email,
			customer_country,
			status,
			created,
			due_by,
		} ) => {
```

File: client/disputes/index.tsx
Comment (mgascam) [reply client/disputes/index.tsx:117]: Happy to change it. It is true that the destructuring assingment is really popular nowadays in the JS world, maybe  too much :)

File: client/disputes/test/__snapshots__/index.tsx.snap
Comment (mgascam) [reply client/disputes/test/__snapshots__/index.tsx.snap:296]: Good point, need to have a look

File: client/disputes/index.tsx
Comment (anu-rock) [reply client/disputes/index.tsx:117]: I think the point about destructuring being memory inefficient is arguable as only a handful of properties are being destructured here. That said, I am in favor of accessing `dispute`s props directly via dot notation for readability.

> It is true that the destructuring assingment is really popular nowadays in the JS world, maybe too much :)

Haha. To a point that a lot of devs no longer care about readability. I once saw the destructuring of 35 or 40 props in a `render` method long enough to fill 20 pages of A4 sheets 😛 After a short while, I lost track of what variable was local and which ones were props.

File: client/disputes/index.tsx
Comment (kalessil) [inline client/disputes/index.tsx:107]: Can we keep eslint enabled, please?

File: client/disputes/style.scss
Comment (kalessil) [inline client/disputes/style.scss:7]: We can omit the comment here.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (mordeth) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: @KarlisJ it was removed by mistake, bringing it back.

Comment (mordeth) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: This block was removed by a mistake, bringing it back.

File: client/types/disputes.d.ts
Comment (kalessil) [inline client/types/disputes.d.ts:16]: Also wondering if we need this definition change.

File: client/types/disputes.d.ts
Comment (kalessil) [inline client/types/disputes.d.ts:105]: I'd expect that count is always presented.

File: client/utils/charge/index.ts
Comment (kalessil) [inline client/utils/charge/index.ts:41]: Not sure if we need changes here at all, my guess is that this was needed due to changes in definitions.

**Review by kalessil — CHANGES_REQUESTED** [review]

File: client/data/disputes/resolvers.js
Comment (mordeth) [reply client/data/disputes/resolvers.js:56]: That code prevents fetching the Dispute object when the details screen is opened. With the previous approach, all the Disputes were preloaded into the store and loaded from there, making the request unnecessary. With the new approach, we no longer fetch the whole Dispute objects from the cached table and the Dispute object needs to be fetched separately when the details page is opened.

File: client/types/charges.d.ts
Comment (mgascam) [reply client/types/charges.d.ts:50]: Good point. It turns out that that now that we are using a dedicated `CachedDisputes` type, we do not interfere with the `Disputes` type. This changes will be reverted

File: client/types/disputes.d.ts
Comment (mgascam) [reply client/types/disputes.d.ts:64]: Not needed anymore, I will revert them

File: client/utils/charge/index.ts
Comment (mgascam) [reply client/utils/charge/index.ts:41]: Yes, not needed, I'll revert them

File: client/types/disputes.d.ts
Comment (mordeth) [reply client/types/disputes.d.ts]: Updated!

File: client/disputes/index.tsx
Comment (mordeth) [reply client/disputes/index.tsx]: Done! :)

File: client/data/disputes/resolvers.js
Comment (kalessil) [reply client/data/disputes/resolvers.js:56]: I see, thank you so much for explaining!

File: client/data/disputes/reducer.js
Comment (kalessil) [inline client/data/disputes/reducer.js:14]: Do we still need `byId` here (also in tests)?

File: client/disputes/test/__snapshots__/index.tsx.snap
Comment (kalessil) [inline client/disputes/test/__snapshots__/index.tsx.snap:322]: Can we update mock, so it matches the original dates, please?

File: client/disputes/test/__snapshots__/index.tsx.snap
Comment (kalessil) [inline client/disputes/test/__snapshots__/index.tsx.snap:345]: Can we also update the mock to match original ids, please?

**Review by kalessil — CHANGES_REQUESTED** [review]
@mordeth, @mgascam - thanks for making new changes, looks great and so much easier to review. Leaving a few more minor comments. Likely we'll need one more final review round and detailed testing and that's it.

File: client/data/disputes/reducer.js
Comment (mordeth) [reply client/data/disputes/reducer.js:14]: No, `byId` is still used as the single Dispute objects are stored in there after fetching.

File: client/disputes/index.tsx
Comment (kalessil) [inline client/disputes/index.tsx:227]: Would `( row[ 3 ].value || '' ) as string` be a good alternative here?

**Review by mgascam — COMMENTED** [review]
LGTM, I did manual testing after latest changes and it is working as expected.

File: client/disputes/details/index.tsx
Comment (kalessil) [inline client/disputes/details/index.tsx:44]: Looks like we could revert this (likely leftovers from types definitions experiments).

**Review by dpaun1985 — APPROVED** [review]
LGTM

**Review by kalessil — COMMENTED** [review]
Added the last 2 comments, let make the changes and I'll start manually testing.

File: client/disputes/index.tsx
Comment (kalessil) [reply client/disputes/index.tsx:227]: Or `( row[ 3 ].value ?? '' ).toString()`

**Review by kalessil — APPROVED** [review]
LGTM! Blocking until Monday, so server changes get deployed first.

**Pull request merged**
